avr-hal-generic/src/serial.rs

```diff
@@ -1,5 +1,75 @@
 //! Serial Implementations
 
+use core::cmp::Ordering;
+
+// Clock is needed because the calculations needs to take core clock into account
+#[derive(Debug, Clone, Copy)]
+pub struct Baudrate<CLOCK> {
+    pub ubrr: u16,
+    pub u2x: bool,
+    pub _clock: ::core::marker::PhantomData<CLOCK>,
+}
+
+impl<CLOCK: crate::clock::Clock> PartialEq for Baudrate<CLOCK> {
+    fn eq(&self, other: &Self) -> bool {
+        self.value() == other.value()
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Eq for Baudrate<CLOCK> {}
+
+impl<CLOCK: crate::clock::Clock> PartialOrd for Baudrate<CLOCK> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.value().cmp(&other.value()))
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.value().cmp(&other.value())
+    }
+}
+
+impl<CLOCK: crate::clock::Clock> Baudrate<CLOCK> {
+    pub fn new(baud: u32) -> Baudrate<CLOCK> {
+        // Try u2x mode first
+        let mut baud_setting = (CLOCK::FREQ / 4 / baud - 1) / 2;
+        let mut u2x = true;
+
+        // Switch back from u2x if required
+        if ((CLOCK::FREQ == 16_000_000) && (baud == 57_600)) || (baud_setting > 4095) {
+            u2x = false;
+            baud_setting = (CLOCK::FREQ / 8 / baud - 1) / 2;
+        }
+
+        let ubrrh = baud_setting >> 8;
+        let ubrrl = baud_setting;
+
+        Baudrate {
+            ubrr: (ubrrh | ubrrl) as u16,
+            u2x: u2x,
+            _clock: ::core::marker::PhantomData,
+        }
+    }
+
+    pub fn with_exact(u2x: bool, ubrr: u16) -> Baudrate<CLOCK> {
+        Baudrate {
+            ubrr: ubrr,
+            u2x,
+            _clock: ::core::marker::PhantomData,
+        }
+    }
+
+    fn value(&self) -> u16 {
+        let divisor = (16 * (self.ubrr + 1)) as u32;
+        if self.u2x {
+            return (CLOCK::FREQ / (divisor / 2)) as u16;
+        } else {
+            return (CLOCK::FREQ / divisor) as u16;
+        };
+    }
+}
```

Comment:
I know what you're trying to do here but this isn't going to work well:  Doing these calculations is expensive and we can actually get around them by comparing inverse values and disregarding the `CLOCK` entirely:

```rust
    fn compare_value(&self) -> u32 {
        if self.u2x {
            8 * (self.ubrr as u32 + 1)
        } else {
            16 * (self.ubrr as u32 + 1)
        }
    }
```

and of course you then need to swap the operands in your `Ord` impl:

```rust
impl<CLOCK: crate::clock::Clock> Ord for Baudrate<CLOCK> {
    fn cmp(&self, other: &Self) -> Ordering {
        // Inverse comparison because compare_value() returns Fosc/BAUD (the inverse value).
        other.compare_value().cmp(&self.compare_value())
    }
}
```

Reply:
Thanks, idea wast to do this 1-2-1 with datasheet and then simplify.